### PR TITLE
(PXP-6693): Fix generation of signed upload urls for public Indexd records for authz-permitted users.

### PR DESCRIPTION
**Pull request by johnfrancismccann**
Jira Ticket: [PXP-6693](https://ctds-planx.atlassian.net/browse/PXP-6693)

With the old logic for generating a signed url for public records, Fence did not allow upload urls to be generated for records with public `authz` or `acl` values ("/open" and "*", respectively). However, users with a `write-storage` permission on "/open" for Fence should be permitted to generate upload urls.

### Bug Fixes
- Allow permitted users to generate a signed upload url for public records (i.e. those records with `authz=[“/open”]`)

### CODE REVIEW

File: fence/blueprints/data/indexd.py [hunk 1]
```diff
@@ -94,9 +94,10 @@ def create_presigned_url_audit_log(indexed_file, action, protocol):
         # fall back on ACL
         resource_paths = indexed_file.index_document.get("acl", [])
     if not protocol:
-        # we can assume there are locations since `get_signed_url()`
-        # used the same logic before this code runs
-        protocol = indexed_file.indexed_file_locations[0].protocol
+        protocol = (
+            indexed_file.indexed_file_locations
+            and indexed_file.indexed_file_locations[0].protocol
+        )
```
Comment (johnfrancismccann) [inline fence/blueprints/data/indexd.py:100]: I added the check because I was getting a 500 error 'list index out of range' for Indexd records with an empty `urls` field (at least that's what I think was the problem). For example, with the old code, setting the `urls` field to `[]` for `INDEXD_RECORD_WITH_PUBLIC_AUTHZ_AND_ACL_POPULATED` and then running `test_public_authz_and_acl_object_upload_file` resulted in a 500 from the list index.

File: fence/blueprints/data/indexd.py [hunk 2]
```diff
@@ -436,8 +454,18 @@ def check_authz(self, action):
...
+
+        header = flask.request.headers.get("Authorization")
+        if header:
+            try:
+                _, token = header.split(" ")
+            except ValueError:
+                raise Unauthorized("authorization header not in expected format")
+        else:
+            token = None
+
```
Comment (johnfrancismccann) [inline fence/blueprints/data/indexd.py:466]: `get_jwt` raises a `401` on a non-existent `Authorization` header, so not using `get_jwt` is to allow for anonymous signed download requests.

File: fence/blueprints/data/indexd.py (hunk 1 above)
Comment (johnfrancismccann) [reply fence/blueprints/data/indexd.py:100]: When `indexed_file.indexed_file_locations` is falsy, it would probably be better to set `protocol` to either `None` or `“”` (instead of an empty list)?

File: fence/blueprints/data/indexd.py [hunk 3]
```diff
@@ -363,14 +364,31 @@ def get_signed_url(
...
+        else:
+            if self.public_acl and action == "upload":
+                raise Unauthorized(
+                    "Cannot upload on public files while using acl field"
+                )
+            # don't check the authorization if the file is public
+            # (downloading public files with no auth is fine)
+            if not self.public_acl and not self.check_authorization(action):
+                raise Unauthorized(
+                    "You don't have access permission on this file: {}".format(
```
Comment (Avantol13) [inline fence/blueprints/data/indexd.py:387]: should use f-string here too, or .format above, up to you but ideally should be consistent

File: fence/blueprints/data/indexd.py (hunk 2 above)
Comment (Avantol13) [reply fence/blueprints/data/indexd.py:466]: can we just try/except instead of implementing custom logic here? might be a bit cleaner

```
try:
  token = get_jwt()
except {{ExceptionObject}} as exc:
  token = None
```

File: fence/blueprints/data/indexd.py [hunk 4]
```diff
@@ -93,9 +93,7 @@ def create_presigned_url_audit_log(indexed_file, action, protocol):
     if not resource_paths:
         # fall back on ACL
         resource_paths = indexed_file.index_document.get("acl", [])
-    if not protocol:
-        # we can assume there are locations since `get_signed_url()`
-        # used the same logic before this code runs
+    if not protocol and indexed_file.indexed_file_locations:
         protocol = indexed_file.indexed_file_locations[0].protocol
```
Comment (johnfrancismccann) [inline fence/blueprints/data/indexd.py:97]: Please see [earlier comment](https://github.com/uc-cdis/fence/pull/893#discussion_r605763295) for why this change was made

File: fence/blueprints/data/indexd.py (hunk 2 above)
Comment (Avantol13) [reply fence/blueprints/data/indexd.py:457]: you may want to add a quick comment for clarication why catching that error is expected and we can continue. something like `raises unauthorized when user is anonymous (no token is available) but we can still evaluate if an anonymous user can access public data, so continue` maybe something less verbose

**Review by Avantol13 — APPROVED** [review]
